src/webui/webapplication.cpp

```diff
@@ -893,98 +831,133 @@ void WebApplication::action_command_setCategory()
+        status(409, "Conflict");
+        print(tr("Incorrect category name"), Http::CONTENT_TYPE_TXT);
         return;
     }
 
     BitTorrent::Session::instance()->addCategory(category);
 }
 
-void WebApplication::action_command_removeCategories()
+void WebApplication::removeCategories()
 {
-    CHECK_URI(0);
     CHECK_PARAMETERS("categories");
 
     QStringList categories = request().posts["categories"].split('\n');
     foreach (const QString &category, categories)
         BitTorrent::Session::instance()->removeCategory(category);
 }
 
-void WebApplication::action_command_getSavePath()
+void WebApplication::getDefaultSavePath()
 {
-    CHECK_URI(0);
     print(BitTorrent::Session::instance()->defaultSavePath());
 }
 
-bool WebApplication::isPublicScope()
+bool WebApplication::isAPIMethod(const QString &methodName) const
 {
-    return (scope_ == DEFAULT_SCOPE || scope_ == VERSION_INFO);
+    return (m_privateAPIs.contains(methodName)
+            || m_publicAPIs.contains(methodName));
+}
+
+bool WebApplication::isPublicAPIMethod(const QString &methodName) const
+{
+    return m_publicAPIs.contains(methodName);
 }
 
 void WebApplication::doProcessRequest()
 {
-    scope_ = DEFAULT_SCOPE;
-    action_ = DEFAULT_ACTION;
+    if (request().path.startsWith(PATH_PREFIX_API)) {
+        const QString methodName {request().path.mid(PATH_PREFIX_API.size())};
+        if (!isAPIMethod(methodName)) {
+            status(404, "Not Found");
+            return;
+        }
 
-    parsePath();
+        if (!sessionActive() && !isPublicAPIMethod(methodName)) {
+            status(403, "Forbidden");
+            return;
+        }
 
-    if (args_.contains(".") || args_.contains("..")) {
-        qDebug() << Q_FUNC_INFO << "Invalid path:" << request().path;
-        status(404, "Not Found");
-        return;
+        QMetaObject::invokeMethod(this, methodName.toLatin1().constData());
     }
-
-    if (!isPublicScope() && !sessionActive()) {
-        status(403, "Forbidden");
-        return;
+    else {
+        sendWebUIFile();
     }
+}
 
-    if (actions_.value(scope_).value(action_) != 0) {
-        (this->*(actions_[scope_][action_]))();
-    }
-    else {
-        status(404, "Not Found");
-        qDebug() << Q_FUNC_INFO << "Resource not found:" << request().path;
+void WebApplication::configure()
+{
+    const auto pref = Preferences::instance();
+    m_username = pref->getWebUiUsername();
+    m_password = pref->getWebUiPassword();
+
+    const QString rootFolder = Utils::Fs::expandPathAbs(
+                !pref->isAltWebUiEnabled() ? WWW_FOLDER : pref->getWebUiRootFolder());
+    if (rootFolder != m_rootFolder) {
+        m_translatedFiles.clear();
+        m_rootFolder = rootFolder;
     }
 }
 
-void WebApplication::parsePath()
+bool WebApplication::sendFile(const QString &path)
 {
-    if (request().path == "/") action_ = WEBUI_ACTION;
+    const QMimeType type {QMimeDatabase().mimeTypeForFile(path, QMimeDatabase::MatchExtension)};
```

Comment:
Now built-in WebUI is broken, FF 57.0b12 initiates a download instead of opening the page.
The returned http content-type is `application/x-extension-html`, weird.

Also tested `QMimeDatabase::MatchDefault` doesn't help, still parsing error in FF.

Reply:
Works fine here.

Reply:
>Works fine here.

Your Qt version?
Qt 5.9.2 here.

Reply:
5.9.2 too, but I think the key component here is the actual mime database. Do you use the bundled one or does your system provide the mime database?

Reply:
5.9.1 works correctly.

Reply:
>Do you use the bundled one or does your system provide the mime database?

I guess the system one. I can see the shared database is installed correctly.
I wonder if this is a package issue...

Reply:
By the way, preventing possible claims at the expense of performance...
>It is perfectly OK to create an instance of QMimeDatabase every time you need to
    perform a lookup.
    The parsing of mimetypes is done on demand (when shared-mime-info is installed)
    or when the very first instance is constructed (when parsing XML files directly).

Reply:
Just noting down what is happening so far:
* with `QMimeDatabase::MatchExtension`, qbt is returning `application/x-extension-html`, so my browser initiates download.
  Weird mime type, don't know why.
* with `QMimeDatabase::MatchDefault`, qbt returns `application/xhtml+xml`, this is correct as the file is in xhtml format.
  But this mime type creates another problem, FF starts using XML parser to parse the HTML file and since the HTML is not well-formed XML, FF is spitting out errors and refuse to render the page.
  Force using `text/html` might workaround this, yet I've not found a way to construct such QMimeType type.

Reply:
Now matches both filename and content. It returns "text/html" for me.

>FF starts using XML parser to parse the HTML file

It's correct behavior for strict XHTML files.
>since the HTML is not well-formed XML, FF is spitting out errors and refuse to render the page.

Strict XHTML should be well-formed XML, but some our Web UI sources aren't. It needs to correct it to follow either XHTML or HTML5 standard.

Reply:
>It's correct behavior for strict XHTML files.

Yeah, from what I've read, it's bad for web development, pages broke too easily.

>It needs to correct it to follow either XHTML or HTML5 standard.

I would chose HTML5.
Also this issue should be solved if QMimeDatabase is being used.